libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1132 @@
+                                  &action_is_filtered_on_this_filter,
+                                  &local_error))
+        {
+          if (g_error_matches (local_error,
+                               EOS_UPDATER_ERROR,
+                               EOS_UPDATER_ERROR_UNKNOWN_ENTRY_IN_AUTOINSTALL_SPEC))
+            {
+              g_warning ("%s. Skipping this filter and not applying action, system may be in an inconsistent state from this point forward", local_error->message);
+              *is_filtered = TRUE;
+              g_clear_error (&local_error);
+              continue;
+            }
+
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return FALSE;
+        }
+
+      if (action_is_filtered_on_this_filter)
+        {
+          *is_filtered = TRUE;
+          return TRUE;
+        }
+    }
+
+  *is_filtered = FALSE;
+  return TRUE;
+}
+
+static GPtrArray *
+read_flatpak_ref_actions_from_file (GFile         *file,
+                                    GCancellable  *cancellable,
+                                    GError       **error)
+{
+  /* Now that we have the file contents, time to read in the list of
+   * flatpaks to install into a pointer array. Parse out the OSTree ref
+   * and then parse the FlatpakRemoteRefAction */
+  g_autoptr(JsonNode) node = parse_json_from_file (file, cancellable, error);
+  g_autoptr(GPtrArray) actions = NULL;
+  JsonArray *array = NULL;
+  g_autoptr(GList) elements = NULL;
+  GList *iter = NULL;
+
+  if (node == NULL)
+    return NULL;
+
+  /* Parse each entry of the underlying array */
+  array = json_node_get_array (node);
+
+  if (array == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (node);
+      g_autofree gchar *filename = g_file_get_path (file);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected node to be an array when parsing %s at %s", node_str, filename);
+
+      return NULL;
+    }
+
+  elements = json_array_get_elements (array);
+  actions = g_ptr_array_new_with_free_func ((GDestroyNotify) flatpak_remote_ref_action_unref);
+
+  for (iter = elements; iter != NULL; iter = iter->next)
+    {
+      /* We use local_error here so that we can catch and
+       * add detail on the filename if necessary */
+      g_autoptr(GError) local_error = NULL;
+      FlatpakRemoteRefAction *action = NULL;
+      gboolean is_filtered = FALSE;
+
+      if (!action_node_should_be_filtered_out (iter->data, &is_filtered, &local_error))
+        {
+          if (g_error_matches (local_error,
+                               EOS_UPDATER_ERROR,
+                               EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC))
+            {
+              g_autofree gchar *filename = g_file_get_path (file);
+              g_propagate_prefixed_error (error, local_error, "Error parsing %s: ", filename);
```

Comment:
You probably need to `return NULL` here, or the error will be overwritten again below.